core/src/Kokkos_ViewHolder.hpp

```diff
@@ -0,0 +1,335 @@
+      view.rank_dynamic > 5 ? view.extent(5) : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+      view.rank_dynamic > 6 ? view.extent(6) : KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+      view.rank_dynamic > 7 ? view.extent(7) : KOKKOS_IMPL_CTOR_DEFAULT_ARG);
+}
+
+class ConstViewHolderImplBase {
+ public:
+  virtual ~ConstViewHolderImplBase() = default;
+
+  size_t span() const { return m_span; }
+  bool span_is_contiguous() const { return m_span_is_contiguous; }
+  const void *data() const { return m_data; }
+  std::string label() const { return m_label; }
+
+  size_t data_type_size() const { return m_data_type_size; }
+  bool is_hostspace() const noexcept { return m_is_hostspace; }
+
+  virtual void deep_copy_to_buffer(unsigned char *buff) = 0;
+  virtual ConstViewHolderImplBase *clone() const        = 0;
+
+ protected:
+  ConstViewHolderImplBase(std::size_t span, bool span_is_contiguous,
+                          const void *data, std::string label,
+                          std::size_t data_type_size, bool is_hostspace)
+      : m_span(span),
+        m_span_is_contiguous(span_is_contiguous),
+        m_data(data),
+        m_label(std::move(label)),
+        m_data_type_size(data_type_size),
+        m_is_hostspace(is_hostspace) {}
+
+ private:
+  size_t m_span             = 0;
+  bool m_span_is_contiguous = false;
+  const void *m_data        = nullptr;
+  std::string m_label;
+  size_t m_data_type_size = 0;
+  bool m_is_hostspace     = false;
+};
+
+class ViewHolderImplBase {
+ public:
+  virtual ~ViewHolderImplBase() = default;
+
+  size_t span() const { return m_span; }
+  bool span_is_contiguous() const { return m_span_is_contiguous; }
+  void *data() const { return m_data; }
+  std::string label() const { return m_label; }
+
+  size_t data_type_size() const { return m_data_type_size; }
+  bool is_hostspace() const noexcept { return m_is_hostspace; }
+
+  virtual void deep_copy_to_buffer(unsigned char *buff)   = 0;
+  virtual void deep_copy_from_buffer(unsigned char *buff) = 0;
+  virtual ViewHolderImplBase *clone() const               = 0;
+
+ protected:
+  ViewHolderImplBase(std::size_t span, bool span_is_contiguous, void *data,
+                     std::string label, std::size_t data_type_size,
+                     bool is_hostspace)
+      : m_span(span),
+        m_span_is_contiguous(span_is_contiguous),
+        m_data(data),
+        m_label(std::move(label)),
+        m_data_type_size(data_type_size),
+        m_is_hostspace(is_hostspace) {}
+
+ private:
+  size_t m_span             = 0;
+  bool m_span_is_contiguous = false;
+  void *m_data              = nullptr;
+  std::string m_label;
+  size_t m_data_type_size = 0;
+  bool m_is_hostspace     = false;
+};
+
+template <typename SrcViewType, typename DstViewType, typename Enabled = void>
+struct ViewHolderImplDeepCopyImpl {
+  static void copy_to_unmanaged(SrcViewType &, void *) {
+    Kokkos::Impl::throw_runtime_exception(
+        "Cannot deep copy a view holder to an incompatible view");
```

Comment:
It's still valid to instantiate a ViewHolder of a type that would cause this to fail (and if dynamic view hooks are enabled a holder will be instantiated for basically every view type, so that is pretty much guaranteed to happen with dynamic view hooks enabled). So we can't actually check at compile time we have to error at runtime if the user actually tries to execute the operation.

We still need compile time checking (the primary template in this case) to avoid generating the deep copy call on invalid types since Kokkos itself will give a compile error if the deep copy is instantiated. I'll add a comment explaining the issue so it's a bit more understandable.